nx/lib/nx/defn/grad.ex

```diff
@@ -446,6 +446,47 @@ defmodule Nx.Defn.Grad do
     [{x, reduce_g(x, opts, g)}]
   end
 
+  defp grad(:product, [x, opts], ans, g) do
+    axes =
+      case opts[:axes] do
+        [] -> Nx.axes(x.shape)
+        nil -> Nx.axes(x.shape)
+        axes -> axes
+      end
+
+    unsqueezed_shape = Enum.reduce(axes, Nx.shape(x), &put_elem(&2, &1, 1))
+    g = Nx.reshape(g, unsqueezed_shape)
+    ans = Nx.reshape(ans, unsqueezed_shape)
+
+    # The derivative of a product with respect to element x_i, is that
+    # product with element x_i removed. Having the total product already
+    # computed, we can divide it by x_i to effectively remove it. This
+    # works as long as x_i is other than 0.
+    #
+    # For products with a single zero element, the derivative with respect
+    # to that particular element is the product of the non-zero elements.
+    #
+    # For products with more zeros, the derivative with respect to any of
+    # the elements is always 0.
+
+    zero? = Nx.equal(x, 0)
+
+    ans_removed_zero =
+      zero?
+      |> Nx.select(1, x)
+      |> Nx.product(axes: axes, keep_axes: true)
+
+    zeros_in_product = Nx.sum(zero?, axes: axes, keep_axes: true)
+    one_zero? = Nx.equal(zeros_in_product, 1)
+    many_zeros? = Nx.greater(zeros_in_product, 1)
+
+    dx = Nx.multiply(g, Nx.divide(ans, x))
```

Comment:
Isn't the division by zero here a problem?

Reply:
Wanted to ask that, I would expect inf/nan that doesn't matter given we replace all the invalid values. But perhaps it's better not to rely on this?

Reply:
Well, it raises for binary backend

```
iex(1)> Nx.divide(Nx.tensor(1), Nx.tensor(0))
** (ArithmeticError) bad argument in arithmetic expression
    (complex 0.4.2) lib/complex.ex:579: Complex.divide/2
    (nx 0.3.0) lib/nx/binary_backend.ex:2371: Nx.BinaryBackend."-binary_to_binary/4-lbc$^1/2-18-"/5
    (nx 0.3.0) lib/nx/binary_backend.ex:662: Nx.BinaryBackend.element_wise_bin_op/4
    iex:1: (file)
```

Though the grad tests pass, so that's a little surprising.

Reply:
Doesn't hurt to use the zeroless tensor since we need it anyway, I will change!